clients/src/main/java/org/apache/kafka/common/protocol/Readable.java

```diff
@@ -53,7 +54,7 @@ default List<RawTaggedField> readUnknownTaggedField(List<RawTaggedField> unknown
         return unknowns;
     }
 
-    default MemoryRecords readRecords(int length) {
+    default BaseRecords readRecords(int length) {
```

Comment:
in kafak serialization, we assume the impl of `BaseRecords` is `MemoryRecords`. For example:
```java
        @Override
        public PartitionProduceData duplicate() {
            PartitionProduceData _duplicate = new PartitionProduceData();
            _duplicate.index = index;
            if (records == null) {
                _duplicate.records = null;
            } else {
                _duplicate.records = MemoryRecords.readableRecords(((MemoryRecords) records).buffer().duplicate());
            }
            return _duplicate;
        }
```

Hence, I'm not sure how this change works if you introduce a non-MemoryRecords impl.

Reply:
> Hence, I'm not sure how this change works if you introduce a non-MemoryRecords impl.

Hard-coded duplicating of `MemoryRecords` is not elegant. So I add a `duplicate` method to BaseRecords. How do you think?